src/HexagonOptimize.cpp

```diff
@@ -414,9 +455,47 @@ class OptimizePatterns : public IRMutator {
     }
 
     void visit(const Cast *op) {
+        // To hit more of the patterns we want, rewrite "double casts"
+        // as two stage casts. This also avoids letting vector casts
+        // fall through to LLVM, which will generate large unoptimized
+        // shuffles.
+        static vector<pair<Expr, Expr>> cast_rewrites = {
+            // Saturating narrowing
+            { u8c(wild_u32x), u8c(u16c(wild_u32x)) },
```

Comment:
Yeah, there's no pattern for some of these. They should just get lowered to vshufe(vmin(vmax(...)))